src/model/scanner.js

```diff
@@ -1,132 +1,122 @@
-const mqtt = require('mqtt');
-const { readMessage } = require('../lib/mqtt_helper');
+const { kdTree } = require('kd-tree-javascript');
+const { Client } = require('./client');
 
-const C2G_BROKER_ADDRESS = 'driver.na.car2go.com';
+const KdTree = kdTree;
 const C2G_VEHICLELIST_TOPIC = 'C2G/S2C/4/VEHICLELIST.GZ';
 const C2G_VEHICLELIST_DELTA_TOPIC = 'C2G/S2C/4/VEHICLELISTDELTA.GZ';
 
+const calcDistance = (a, b) => {
+  const rad = Math.PI / 180;
+  let lat1 = a.latiitude;
+  const lon1 = a.longitude;
+  let lat2 = b.latitude;
+  const lon2 = b.longitude;
+  const dLat = (lat2 - lat1) * rad;
+  const dLon = (lon2 - lon1) * rad;
+  lat1 *= rad;
+  lat2 *= rad;
+  const x = Math.sin(dLat / 2);
+  const y = Math.sin(dLon / 2);
+  const dist = x * x + y * y * Math.cos(lat1) * Math.cos(lat2);
+  return Math.atan2(Math.sqrt(dist), Math.sqrt(1 - dist));
+};
+
+const _log = (msg) => {
+  console.log(`[Scanner] – ${msg}`);
+};
 class Scanner {
   constructor(account) {
-    this.account = account;
+    const scannerClient = new Client(account);
+    this.scannerClient = scannerClient;
+    this.tree = null;
+    this.treeMap = {};
+
+    scannerClient.setConnectCallback(
+      () => {
+        scannerClient.subscribe(C2G_VEHICLELIST_TOPIC, 0, this._onReceiveVehicleList.bind(this));
+        scannerClient.subscribe(C2G_VEHICLELIST_DELTA_TOPIC, 1, this._onReceiveVehicleDelta.bind(this));
+      },
+    );
+
+    this.scannerClient.connect();
+
+    this.clients = new Map(); // Client => location
   }
 
-  /**
-   * Connect to Car2Go broker server using client certificate
-   * @param {*} certificate
-   */
-  connect(certificate) {
-    this.client = mqtt.connect({
-      host: C2G_BROKER_ADDRESS,
-      port: 443,
-      protocol: 'mqtts',
-      ca: certificate,
-      clientId: this.account.clientId,
-      username: this.account.clientId,
-      password: this.account.accessToken,
-      protocolVersion: 3,
-      protocolId: 'MQIsdp',
-      rejectUnauthorized: true,
+  _onReceiveVehicleList(data) {
+    this.scannerClient.unsubscribe(C2G_VEHICLELIST_TOPIC);
+    const points = [];
+
+    // initial build
+    data.connectedVehicles.forEach((vehicle) => {
+      points.push(this._buildPointFromVehicle(vehicle));
     });
 
-    this.client.on('connect', () => {
-      this._log('Client connected.');
-      this.client.subscribe(C2G_VEHICLELIST_TOPIC, { qos: 0 }, () => {
-      });
+    this.tree = new KdTree(points, calcDistance, ['latitude', 'longitude']);
+  }
 
-      this.client.subscribe(C2G_VEHICLELIST_DELTA_TOPIC, { qos: 1 }, () => {
-      });
+  _buildPointFromVehicle(vehicle) {
+    const vehicleData = {
```

Comment:
use object destructuring for id, latitude, and longitude.

e.g.

```suggestion
    const { 
        id, geoCoordinate: {
            latitude,
            longitude,
         } = vehicle;
    
```